### PR TITLE
Table Theme, Badge, Table Badge, Table Text

### PR DESCRIPTION
**Pull request by hchamorro**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Enhanced `Badge` component with new props and styles.
- Added `TableBadge` component with editing capabilities and updated stories.
- Implemented `TableTextField` component with support for editing and variants.
- Updated various table components to use the new table theme.
- Added new demo configurations and stories for `TableBadge` and `TableTextField`.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>15 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>Badge.stories.tsx</strong><dd><code>Update Badge stories to include iconName prop</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/demo/stories/Badge.stories.tsx

<li>Added <code>iconName</code> prop to <code>Badge</code> component in stories.<br> <li> Changed layout to center align items.<br>

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-14ac708a02bd45bb565cf8d8f8e879a5bfab4d1270660b3951f21717f32a4680">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>TableBadge.stories.tsx</strong><dd><code>Enhance TableBadge stories with headings and editing options</code></dd></summary>
<hr>

apps/demo/stories/TableBadge.stories.tsx

<li>Added <code>Heading</code> components for better sectioning.<br> <li> Updated <code>TableBadge</code> component props and layout.<br> <li> Added editing options for <code>TableBadge</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-267523eee163711b1922c1675af79f700a962e8eaf8338935dd4828677803d7a">+22/-11</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>TableTextField.stories.tsx</strong><dd><code>Add TableTextField component demo and variants</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/demo/stories/TableTextField.stories.tsx

<li>Added <code>TableTextField</code> component demo.<br> <li> Added variants for <code>TableTextField</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-2ab73d8cc798f678078ad433f2011785aebddf9faa7c1467864690cb666f0a0d">+29/-36</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Enable TableBadge and TableTextField stories</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/demo/stories/index.ts

<li>Un-commented exports for <code>TableBadge</code> and <code>TableTextField</code> stories.<br>

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-590db40ce1849d210cdc32b50de4ad53b837bf58ff0e88d843cb1ff560855fdd">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>TableBadge.config.tsx</strong><dd><code>Add controls to TableBadge demo configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/demo/story-config/TableBadge.config.tsx

- Added controls for `TableBadge` demo configuration.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-45cd7da37a6cfd2d891cbd6add34260b29a5bdd4ba1cc8ad8873207354f19328">+13/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>TableTextField.config.tsx</strong><dd><code>Update TableTextField demo configuration with variants</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/demo/story-config/TableTextField.config.tsx

- Updated `TableTextField` demo configuration to include variants.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-e5ce89f62d7b4977c4e201bb24aa1f334033b4b433039e24db892c80fd0931e6">+7/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Badge.tsx</strong><dd><code>Refactor Badge component with new props and styles</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/ui/src/Badge.tsx

<li>Refactored <code>Badge</code> component to support new props and styles.<br> <li> Added secondary border colors and improved layout.<br>

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-7df4a7c83b27a544f119217086fa0450a34556056feceae8204da72453847a59">+49/-28</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Common.ts</strong><dd><code>Extend Common interfaces for Table and Badge components</code>&nbsp; &nbsp; </dd></summary>
<hr>

packages/ui/src/Common.ts

<li>Added <code>TableTheme</code> interface.<br> <li> Added new props for <code>Badge</code>, <code>TableBadge</code>, and <code>TableTextField</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-538e93b553971318d69ff6fff2ac103f295c4922e24bbe61c30897baa463bba1">+96/-7</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>PickerSelect.tsx</strong><dd><code>Make placeholder prop optional in RNPickerSelect</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/ui/src/PickerSelect.tsx

- Made `placeholder` prop optional in `RNPickerSelect`.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-676b8baa3c288b46a830357bdeb80853e12bdd7e1e7d1d8e60d6a8a3c2c1b227">+1/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Theme.tsx</strong><dd><code>Add default table theme properties to theme configuration</code></dd></summary>
<hr>

packages/ui/src/Theme.tsx

<li>Added default table theme properties to <code>defaultPrimitives</code>.<br> <li> Included <code>table</code> theme in <code>defaultTheme</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-3b9c81e39312b8182e4893dad6e338961ef421fbab9ae28bb1a1edf79367cbab">+16/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>TableBadge.tsx</strong><dd><code>Implement TableBadge component with editing capabilities</code>&nbsp; </dd></summary>
<hr>

packages/ui/src/table/TableBadge.tsx

- Implemented `TableBadge` component with editing capabilities.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-71a8b1708d17b032e9f47d44da7b128637abff884c3e5381ec4bf9dbcff4063c">+59/-15</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>TableBoolean.tsx</strong><dd><code>Update TableBoolean component to use table theme</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/ui/src/table/TableBoolean.tsx

- Updated `TableBoolean` component to use table theme for max width.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-28cb0dfda3909f10b0f64c5cf32d9f84e441fba3381a4e44b54d83f349f26009">+16/-2</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>TableDate.tsx</strong><dd><code>Update TableDate component to use table theme</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/ui/src/table/TableDate.tsx

<li>Updated <code>TableDate</code> component to use table theme for max width and <br>padding.<br>

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-5207f5049c7a8bd8f7b6903ba21d4da28657e274c216f2ba7fed9ed065f41b97">+10/-6</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>TableIconButton.tsx</strong><dd><code>Update TableIconButton component to use table theme</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/ui/src/table/TableIconButton.tsx

<li>Updated <code>TableIconButton</code> component to use table theme for max width.<br>

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-74e3be7fccc4e80eca0e9b22465bcbeadd273fa16003d131ae8e356dba4caa4b">+18/-9</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>TableTextField.tsx</strong><dd><code>Implement TableTextField component with editing and variants</code></dd></summary>
<hr>

packages/ui/src/table/TableTextField.tsx

<li>Implemented <code>TableTextField</code> component with support for editing and <br>variants.<br>

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/615/files#diff-030c3703fc4a3360d7876253dd0604c2f05b4d3aca12e94e7233f02ae4492049">+57/-5</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Review by joshgachnang — COMMENTED** [review]
Reviewed, updated, added code!

**Comment by joshgachnang** [comment]
FYI, there's still some stuff related to editing tables in here, even though we don't support that yet. Leaving it in for now since we'll want to put that in soon.

File: packages/ui/src/table/TableBadge.tsx [hunk 1]
```diff
@@ -1,19 +1,53 @@
...
-};
+// TODO: Support error state in TableBadge
+export const TableBadge = forwardRef<TableBadgeHandles, TableBadgeProps>(
+  (
+    {value, badgeStatus = "info", badgeIconName, isEditing = false, editingOptions, onSave},
+    ref
+  ) => {
+    const [selected, setSelected] = useState<string | undefined>(value);
+
+    useImperativeHandle(ref, () => ({
```
Comment (swanka) [inline packages/ui/src/table/TableBadge.tsx:20]: I'm learning a new hook here. Why do we need to use the imperative handle here if there's no underlying DOM node that we're ref-ing?

File: packages/ui/src/table/TableHeaderCell.tsx [hunk 2]
```diff
@@ -12,58 +16,78 @@ export const TableHeaderCell = ({
...
-      <Box flex="grow" justifyContent="center" marginBottom={2} marginTop={2} width={width}>
-        {children}
-      </Box>
-    );
+  if (sortable && !onSortChange) {
+    console.error("onSortChange is required when sortable is true");
   }
+  return (
+    <Box
+      // accessibilityHint="press to change sorting alphabetical order"
```
Comment (swanka) [inline packages/ui/src/table/TableHeaderCell.tsx:59]: I'm guessing this is waiting on the BoxProps change you made for accessibility?

File: packages/ui/src/table/TableBadge.tsx (hunk 1 above)
Comment (joshgachnang) [reply packages/ui/src/table/TableBadge.tsx:20]: I'm not entirely sure, I think this was going to be related to saving. I can remove it.

File: packages/ui/src/table/TableHeaderCell.tsx (hunk 2 above)
Comment (joshgachnang) [reply packages/ui/src/table/TableHeaderCell.tsx:59]: Oh yes! Updated!

File: packages/ui/src/Badge.tsx [hunk 3]
```diff
@@ -1,23 +1,31 @@
...
-  variant = "text",
+  variant,
   maxValue = 100,
 }: BadgeProps): React.ReactElement => {
   const {theme} = useContext(ThemeContext);
+  const isIconOnly = variant === "iconOnly";
 
   let badgeColor: keyof TextTheme = "inverted";
 
+  const secondaryBorderColors = {
```
Comment (kjkurtz) [inline packages/ui/src/Badge.tsx:21]: No way to get these from the theme?

**Review by kjkurtz — COMMENTED** [review]
Something is a bit goofy with how it renders the table on iOS, but overall styles look good. It blows up when I try to load it on Android though so couldn't test there.

File: packages/ui/src/Badge.tsx (hunk 3 above)
Comment (joshgachnang) [reply packages/ui/src/Badge.tsx:21]: Looks like they're not theme colors in Figma, except that doesn't make sense if someone's going to change the theme. I'll add a todo (and I'll chat with Jo)

**Comment by joshgachnang** [comment]
@kjkurtz ok, table should be fixed up on mobile, works on Android too

**Review by kjkurtz — APPROVED** [review]
Works now!

**Pull request merged**